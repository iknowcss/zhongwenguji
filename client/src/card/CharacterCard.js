import React from 'react';
import cx from 'classnames';
import formatPinyin from '../util/formatPinyin';
import style from './CharacterCard.module.scss'

const WIDEN_SLASH_REGEX = /\//g;

function formatPinyinArray(pinyin) {
  return pinyin.map(formatPinyin).join(', ');
}

function formatDefinitionArray(definition) {
  if (!definition || definition.length === 0) {
    return '(No definition available)';
  }
  return definition.map(d => d.replace(WIDEN_SLASH_REGEX, ' / ')).join(' / ');
}

export default (props) => {
  const {
    showDefinition,
    character,
    pinyin,
    definition,
    score,
    className
  } = props;

  const finalClassName = cx(style.container, {
    [style.showDefinition]: showDefinition,
    [style.markedUnknown]: score === 0,
    [style.markedKnown]: score === 1
  }, className);

  return (
    <span className={finalClassName}>
      {showDefinition ? (
        <>
          <span className={style.pinyin}>
            {formatPinyinArray(pinyin)}
          </span>
          <span className={style.definition}>
            {formatDefinitionArray(definition)}
          </span>
        </>
      ) : (
        <span className={style.character}>
          {character}
        </span>
      )}
    </span>
  )
};
