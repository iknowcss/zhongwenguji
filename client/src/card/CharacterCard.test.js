import React from 'react';
import renderer from 'react-test-renderer';
import CharacterCard from './CharacterCard';

describe('CharacterCard', () => {
  let component;

  function setup(overrides) {
    const props = {
      i: 33,
      cs: '对',
      ct: '對',
      p: ['dui4'],
      d: ['right/correct']
    };

    component = renderer.create(
      <CharacterCard {...props} {...overrides} />
    );
  }

  it('renders the character', () => {
    setup();
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders the definition', () => {
    setup({ showDefinition: true });
    const tree = component.toJSON();

    expect(tree.props.className.trim().split(/\s+/))
      .toEqual(['container', 'showDefinition']);
    expect(tree).toMatchSnapshot();
  });

  it('renders the fallback definition', () => {
    setup({ showDefinition: true, d: undefined });
    const tree = component.toJSON();

    expect(tree.props.className.trim().split(/\s+/))
      .toEqual(['container', 'showDefinition']);
    expect(tree).toMatchSnapshot();
  });

  it('has a class for unknown marking', () => {
    setup({ known: false });
    const tree = component.toJSON();

    expect(tree.props.className.trim().split(/\s+/))
      .toEqual(['container', 'markedUnknown']);
  });

  it('has a class for unknown marking', () => {
    setup({ known: true });
    const tree = component.toJSON();

    expect(tree.props.className.trim().split(/\s+/))
      .toEqual(['container', 'markedKnown']);
  });
});
