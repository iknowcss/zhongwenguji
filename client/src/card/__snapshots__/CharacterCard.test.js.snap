// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterCard renders the character 1`] = `
<span
  className="container"
>
  <span
    className="character"
  >
    对
  </span>
</span>
`;

exports[`CharacterCard renders the definition 1`] = `
<span
  className="container showDefinition"
>
  <span
    className="pinyin"
  >
    duì
  </span>
  <span
    className="definition"
  >
    right / correct
  </span>
</span>
`;

exports[`CharacterCard renders the fallback definition 1`] = `
<span
  className="container showDefinition"
>
  <span
    className="pinyin"
  >
    duì
  </span>
  <span
    className="definition"
  >
    (No definition available)
  </span>
</span>
`;
